packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ContainerRuntimeMessage } from "..";
+import { IBatch } from "./definitions";
+
+interface IGroupedMessage {
+	contents: string;
+	metadata: Record<string, unknown> | undefined;
+	compression: string;
+}
+
+export class OpGroupingManager {
+	static groupedBatchOp = "groupedBatch";
+
+	constructor(private readonly groupedBatchingEnabled: boolean) {}
+
+	public groupBatch(batch: IBatch): IBatch {
+		if (batch.content.length < 2 || !this.groupedBatchingEnabled) {
+			return batch;
+		}
+
+		// Need deserializedContent for back-compat
+		const deserializedContent = {
+			type: OpGroupingManager.groupedBatchOp,
+			contents: batch.content.map((message) => ({
+				contents: message.contents,
+				metadata: message.metadata,
```

Comment:
I left a comment elsewhere about it, but this code should assert that there is no metadata on op evenlope. metadata at that level is supposed to be inspectable by service, and grouping kills that capability. Two examples I know (blobId and compression) are an examles where something gets broken if metadata is here

Compression is worth separate look: Initially we wanted compression to be on op envelope for service to be able to re-pack ops if it needs to (for example, for long term storage after they were acked and arrive to SPO). But with chunking, this is not possible. I think chunking code should reflect it - once chunking is enabled, it should strip op envelope from compression metadata, and keep it as part of content.
We could re-evaluate this whole approach of putting it on op envelope (if most compressed payloads ends up chunked, then not much value to sometimes give a chance for service to observe compression).